### PR TITLE
feat(metrics): RHICOMPL-2316 LRU cache dangling items

### PR DESCRIPTION
**Pull request by vkrizan**
Adds LRU cache on engineering prometheus collector, that mostly
collects dangling items.

This should alleviate hits to database, and should fix database high
load.

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [X] Access Control
- [X] Cryptographic Practices
- [X] Error Handling and Logging
- [X] Data Protection
- [X] Communication Security
- [X] System Configuration
- [X] Database Security
- [X] File Management
- [X] Memory Management
- [X] General Coding Practices

### CODE REVIEW

File: lib/prometheus/engineering_collector.rb [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 
+require 'lru_redux'
+
 unless defined? Rails
   require File.expand_path('../../config/environment', __dir__)
 end
 
 # when /metrics is called. This runs directly on prometheus_exporter.
 class EngineeringCollector < PrometheusExporter::Server::TypeCollector
+  LRU_MINUTES = 15
```
Comment (skateman) [inline lib/prometheus/engineering_collector.rb:11]: I think for these specific metrics below the number can be much higher. I would go with 60 minutes, but having like 12 hours is also okay from the data integrity side. On the other hand, why should we keep something in-memory for days :shrug:

Comment (vkrizan) [reply lib/prometheus/engineering_collector.rb]: One hour it is.

**Review by skateman — APPROVED** [review]
The Seal of Approval
Looks good to me, works on my machine...but it probably needs the stress of the stage environment to see if it's working for real

File: lib/prometheus/engineering_collector.rb [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 unless defined? Rails
   require File.expand_path('../../config/environment', __dir__)
 end
 
 # when /metrics is called. This runs directly on prometheus_exporter.
 class EngineeringCollector < PrometheusExporter::Server::TypeCollector
+  LRU_DURATION = 1.hour
+
   def initialize
+    @cache = LruRedux::TTL::ThreadSafeCache.new(1, LRU_DURATION.seconds.to_i)
```
Comment (akofink) [inline lib/prometheus/engineering_collector.rb:14]: Why introduce a new dependency when we could use redis with Rails.cache?

Comment (vkrizan) [reply lib/prometheus/engineering_collector.rb:14]: Can we specify timout with generic Rails caching?

Comment (skateman) [reply lib/prometheus/engineering_collector.rb:14]: If not, we can rely on redis' own methods, there you can set a TTL.

Comment (akofink) [reply lib/prometheus/engineering_collector.rb:14]: Yes, there is an `expires_in` option. See https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch

**Comment by akofink** [comment]
I hope we're viewing this as a band-aid and not a complete solution. Metrics per pod greatly increases the load on the DB for no benefit to us (backend and sidekiq both run with 6 pods each, for example). I'd like to figure out how we can re-architect our metrics rather than working around the poor existing architecture. Also, our metrics are broken right now in production due similarly to the poor architecture we've landed on.

**Comment by vkrizan** [comment]
@akofink Indeed this is a kind of a band-aid. The LRU cache will likely stay, however the architecture would need some thought (as you've mentioned). Pods should export only pod-specific metrics. Extra metrics would likely need its own service/pod.

**Comment by vkrizan** [comment]
@akofink Indeed this is a kind of a band-aid. The LRU cache will likely stay, however the architecture would need some thought (as you've mentioned). Pods should export only pod-specific metrics. Extra metrics would likely need its own service/pod.

File: lib/prometheus/engineering_collector.rb (hunk 2 above)
Comment (akofink) [reply lib/prometheus/engineering_collector.rb:14]: I am opposed to adding a new unnecessary dependency. Looking at the gem, it doesn't even appear active (last changes were in 2017; last release was March 2015) or well documented (docs link broken: https://www.rubydoc.info/gems/lru_redux/1.1.0). We need to be careful when selecting new dependencies and ideally not introduce them at all where not necessary.

Comment (vkrizan) [reply lib/prometheus/engineering_collector.rb:14]: Yes, I'm on the same page. The lib was recommended by Racecar. I've checked activeness of that project and it seemed fine (code is stable and there are discussions happening).

**Comment by vkrizan** [comment]
@akofink @skateman PTAL

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Comment by akofink** [comment]
Nice! Thanks, @vkrizan! Maybe we should also do the same for the business_collector.

**Pull request merged**